modules/src/ics02_client/context.rs

```diff
@@ -15,25 +14,28 @@ pub trait ClientReader {
     fn client_type(&self, client_id: &ClientId) -> Option<ClientType>;
     fn client_state(&self, client_id: &ClientId) -> Option<AnyClientState>;
     fn consensus_state(&self, client_id: &ClientId, height: Height) -> Option<AnyConsensusState>;
+
+    /// Returns a natural number, counting how many clients have been created thus far.
+    /// The value of this counter should increase only via method `ClientKeeper::increase_client_counter`.
+    fn client_counter(&self) -> u64;
 }
 
 /// Defines the write-only part of ICS2 (client functions) context.
 pub trait ClientKeeper {
-    fn store_client_result(
-        &mut self,
-        handler_res: ClientResult,
-    ) -> Result<Vec<ClientEvent>, Error> {
+    fn store_client_result(&mut self, handler_res: ClientResult) -> Result<(), Error> {
         match handler_res {
             Create(res) => {
-                let client_id = self.next_client_id();
+                let client_id = res.client_id.clone();
+
                 self.store_client_type(client_id.clone(), res.client_type)?;
                 self.store_client_state(client_id.clone(), res.client_state.clone())?;
                 self.store_consensus_state(
-                    client_id.clone(),
+                    client_id,
                     res.client_state.latest_height(),
                     res.consensus_state,
                 )?;
-                Ok(vec![ClientEvent::ClientCreated(client_id)])
```

Comment:
Events were being created both here and at the handlers. Now they're only created at the handlers.